File: solutions/00003/python/main.py
import math

def int_sqrt(n: int) -> int:
    return math.floor(math.sqrt(n))

def is_prime(n: int):
    if n % 2 == 0:
        return False
    for i in range(3, int_sqrt(n) + 1, 2):
        if n % i == 0:
            return False
    return True

INPUT = 600851475143
# INPUT = 13195

if __name__ == "__main__":
    for a in range(int_sqrt(INPUT) - 1, 3, -2):
        if INPUT % a == 0 and is_prime(a):
            print(a)

File: solutions/00003/clojure/main.clj
(ns main
  (:require [clojure.math :as math]))

(defn int-sqrt [x]
  (math/floor (math/sqrt x)))

(defn prime? [x]
  (let [nums (cons 2 (range 3 (int-sqrt x) 2))
        mods (map #(mod x %) nums)]
    (not (some zero? mods))))

(defn prime-factors
  ([x] (prime-factors x (if (even? x) (dec (/ x 2)) (dec (/ (dec x) 2)))))
  ([x n] (if (zero? n)
           '(0)
           (if (and (zero? (mod x n)) (prime? x))
             (lazy-seq (cons x (prime-factors x (- n 2))))
             (lazy-seq (prime-factors x (- n 2)))))))

(println (prime-factors 13195))
